File: python/open_redirect/account_management.py
from flask import Flask, request, redirect, url_for

app = Flask(__name__)


# VULN 1: Open redirect on password reset - redirect_to param not validated
@app.route("/password-reset/confirm")
def password_reset_confirm():
    token = request.args.get("token")
    redirect_to = request.args.get("redirect_to", "/account")
    if validate_reset_token(token):
        return redirect(redirect_to)
    return "Invalid token", 400


# VULN 2: Open redirect in referrer-based navigation - Referer header used directly
@app.route("/logout")
def logout():
    referer = request.headers.get("Referer", "/")
    # clear session...
    return redirect(referer)


# VULN 3: Open redirect in account linking - callback_url for third-party accounts
@app.route("/account/link-social")
def link_social_account():
    provider = request.args.get("provider")
    callback_url = request.args.get("callback_url")
    # link provider...
    return redirect(callback_url)


def validate_reset_token(token):
    return True

File: javascript/open_redirect/accountManagement.js
const express = require('express');

const router = express.Router();

// VULN 1: Unvalidated 'redirect_to' param in password reset confirmation
router.post('/password-reset/confirm', (req, res) => {
    const { token, newPassword, redirect_to } = req.body;
    const isValid = validateResetToken(token);
    if (!isValid) return res.status(400).json({ error: 'Invalid or expired reset token' });
    updateUserPassword(token, newPassword);
    // redirect_to not validated against an allowlist - open redirect after password reset
    res.redirect(redirect_to || '/account/login');
});

// VULN 2: Referer header used directly in logout redirect
router.post('/logout', (req, res) => {
    const referer = req.headers['referer'];
    req.session.destroy();
    // Referer header is attacker-controllable - redirects to forged origin after logout
    res.redirect(referer || '/');
});

// VULN 3: Unvalidated 'callback_url' for social account linking
router.get('/account/social/link', (req, res) => {
    const { provider, callback_url } = req.query;
    const authState = generateOAuthState(req.session.userId, callback_url);
    const oauthUrl = buildOAuthUrl(provider, authState);
    // callback_url stored in state and later used for redirect without domain validation
    res.redirect(oauthUrl);
});

router.get('/account/social/callback', (req, res) => {
    const { code, state } = req.query;
    const { userId, callback_url } = parseOAuthState(state);
    linkSocialAccount(userId, code);
    // callback_url from state is user-supplied - redirects to arbitrary URL after linking
    res.redirect(callback_url || '/account/settings');
});

function validateResetToken(token) { return true; }
function updateUserPassword(token, password) {}
function generateOAuthState(userId, callbackUrl) { return Buffer.from(JSON.stringify({ userId, callback_url: callbackUrl })).toString('base64'); }
function buildOAuthUrl(provider, state) { return `https://oauth.${provider}.com/auth?state=${state}`; }
function parseOAuthState(state) { return JSON.parse(Buffer.from(state, 'base64').toString()); }
function linkSocialAccount(userId, code) {}

module.exports = router;
